dashboard/src/controllers/widgets/user-role-widget-controller.js

```diff
@@ -1,325 +1,385 @@
+        super("user-role-selector.html", () => {
+            onUserRoleSave(this._getUsersRoles(), this._getCustomSettings());
+        }, () => {
+            this._guild = guild;
+            this._headerText = headerText;
+
+            this._usersRoles = usersRoles.map((userRole) => {
+                return new Discotron.UserRole(userRole.discordUserId, userRole.discordRoleId);
+
+            });
+
+            this._displayRoles = displayRoles;
+            this._allowNone = allowNone;
+            this._customInputs = customInputs;
+
+            document.querySelector("#users-roles-container").style.display = "none";
+            document.querySelector("#input-description").textContent = inputHelp;
+            document.querySelector("#add-button").disabled = true;
+
+            this._displayCustomElements();
+            this._displayUserRoleSelector();
+        }, onClose);
+
+    }
+
+    /**
+     * @returns {array} List of userRoles selected by the user
+     */
+    _getUsersRoles() {
+        return this._usersRoles;
+    }
+
+    /**
+     * @returns {object} Custom settings as key -> value
+     */
+    _getCustomSettings() {
+        // Switches
+        let customSettings = {};
+
+        let switches = this._widgetContainer.querySelectorAll(".custom-switch-checkbox");
+        for (let i = 0; i < switches.length; i++) {
+            const checkbox = switches[i];
+            customSettings[checkbox.dataset.devName] = checkbox.checked;
+        }
+
+        return customSettings;
+    }
+
+
+    /**
+     * Handle search selector helper and more
+     */
+    _addEvents() {
+        super._addEvents();
+
+        let input = document.getElementById("name-input");
+        let button = document.getElementById("add-button");
+
+        input.oninput = () => {
+            if (input.value !== "") {
+                this._checkNameValidity(input.value);
+            }
+        };
+
+        input.onpaste = (event) => {
+            let paste = (event.clipboardData || window.clipboardData).getData("text");
+            this._checkIdValidity(paste);
+            event.preventDefault();
+        };
+
+        input.onkeydown = (e) => {
+            if (e.keyCode === 13) {
+                button.click();
+            }
+        };
+
+        button.onclick = () => {
+            button.disabled = true;
+            let value = input.value;
+            input.value = "";
+
+            // The text is the name of a role or a user
+            this._addEntry(value);
+        };
+    }
+
+    /**
+     * Activates the add button if the name corresponds to a valid discord name, or a valid user id
+     * @param {string} name Username of a discord user or role name
+     */
+    _checkIdValidity(id) {
+        let button = document.querySelector("#add-button");
+        button.disabled = true;
+
+        // This was not an existing user, check if it's an id
+        if (id.match(/^[0-9]*$/)) {
```

Comment:
Replace * with +, else an empty discord id could *in theory* be accepted by this function.